pkg/sql/catalog/tabledesc/constraint.go

```diff
@@ -11,215 +11,477 @@
+}
+
+// OnDelete implements the catalog.ForeignKeyConstraint interface.
+func (c foreignKeyConstraint) OnDelete() catpb.ForeignKeyAction {
+	return c.desc.OnDelete
+}
+
+// OnUpdate implements the catalog.ForeignKeyConstraint interface.
+func (c foreignKeyConstraint) OnUpdate() catpb.ForeignKeyAction {
+	return c.desc.OnUpdate
+}
+
+// Match implements the catalog.ForeignKeyConstraint interface.
+func (c foreignKeyConstraint) Match() descpb.ForeignKeyReference_Match {
+	return c.desc.Match
+}
+
+// GetConstraintID implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) GetConstraintID() descpb.ConstraintID {
+	return c.desc.ConstraintID
+}
+
+// GetConstraintValidity implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) GetConstraintValidity() descpb.ConstraintValidity {
+	return c.desc.Validity
+}
+
+// IsConstraintValidated implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) IsConstraintValidated() bool {
+	return c.desc.Validity == descpb.ConstraintValidity_Validated
+}
+
+// IsConstraintUnvalidated implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) IsConstraintUnvalidated() bool {
+	return c.desc.Validity == descpb.ConstraintValidity_Unvalidated
+}
+
+// GetName implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) GetName() string {
+	return c.desc.Name
+}
+
+// AsForeignKey implements the catalog.ConstraintProvider interface.
+func (c foreignKeyConstraint) AsForeignKey() catalog.ForeignKeyConstraint {
+	return &c
+}
+
+// String implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) String() string {
+	return fmt.Sprintf("%+v", c.desc)
+}
+
+// IsEnforced implements the catalog.Constraint interface.
+func (c foreignKeyConstraint) IsEnforced() bool {
+	return !c.IsMutation() || c.WriteAndDeleteOnly()
+}
+
+// constraintCache contains precomputed slices of constraints, categorized by:
+//   - constraint subtype: checks, fks, etc.
+//   - enforcement status: whether a constraint is enforced for data written to
+//     the table, regardless of whether the constraint is valid for table data
+//     which existed prior to the constraint being added to the table.
+type constraintCache struct {
+	all, allEnforced       []catalog.Constraint
+	checks, checksEnforced []catalog.CheckConstraint
+	fks, fksEnforced       []catalog.ForeignKeyConstraint
+	uwis, uwisEnforced     []catalog.UniqueWithIndexConstraint
+	uwois, uwoisEnforced   []catalog.UniqueWithoutIndexConstraint
+	fkBackRefs             []catalog.ForeignKeyConstraint
+}
+
+// newConstraintCache returns a fresh fully-populated constraintCache struct for the
+// TableDescriptor.
+func newConstraintCache(
+	desc *descpb.TableDescriptor, indexes *indexCache, mutations *mutationCache,
+) *constraintCache {
+	capUWIs := len(indexes.all)
+	numEnforcedChecks := len(desc.Checks)
+	capChecks := numEnforcedChecks + len(mutations.checks)
```

Comment:
nit: the naming seems to suggest all `desc.Checks` are enforced checks and all `mutations.checks` are not enforced checks, which isn't the case, right?

A check on the mutation with VALIDATING validity is considered enforced, right?

Reply:
See above.